build-scripts/k8s_release.py

```diff
@@ -18,72 +18,131 @@
+    """
+    for tag in get_k8s_tags():
         if is_stable_release(tag):
             return tag
-    raise ValueError("Couldn't find stable release, received tags: %s" % k8s_tags)
+    raise ValueError("Couldn't find a stable release.")
 
 
-def get_latest_release() -> str:
-    k8s_tags = get_k8s_tags()
-    return k8s_tags[0]
+def get_latest_releases_by_minor() -> Dict[str, str]:
+    """Map each minor Kubernetes version to its latest release tag.
+
+    Returns:
+        A dictionary mapping minor versions (e.g. '1.30') to the
+        latest (pre-)release tag (e.g. 'v1.30.1').
+    """
+    latest_by_minor: Dict[str, str] = {}
+    version_regex = re.compile(r"^v?(\d+)\.(\d+)\..+")
+
+    for tag in get_k8s_tags():
+        match = version_regex.match(tag)
+        if not match:
+            continue
+        major, minor = match.groups()
+        key = f"{major}.{minor}"
+        if key not in latest_by_minor:
+            latest_by_minor[key] = tag
 
+    return latest_by_minor
 
-def get_outstanding_prerelease() -> Optional[str]:
-    latest_release = get_latest_release()
-    if not is_stable_release(latest_release):
-        return latest_release
-    # The latest release is a stable release, no outstanding pre-release.
-    return None
+
+def get_outstanding_prereleases(as_git_branch: bool = False) -> List[str]:
+    """Return outstanding K8s pre-releases.
+
+    Args:
+        as_git_branch: If True, return the git branch name for the pre-release.
+    """
+    latest_release = get_latest_releases_by_minor()
+    prereleases = []
+    for tag in latest_release.values():
+        if not is_stable_release(tag):
+            prereleases.append(tag)
+
+    if as_git_branch:
+        return [get_prerelease_git_branch(tag) for tag in prereleases]
+
+    return prereleases
 
 
 def get_obsolete_prereleases() -> List[str]:
     """Return obsolete K8s pre-releases.
 
-    We only keep the latest pre-release if there is no corresponding stable
+    We only keep the latest pre-release(s) if there is no corresponding stable
     release. All previous pre-releases are discarded.
     """
     k8s_tags = get_k8s_tags()
-    if not is_stable_release(k8s_tags[0]):
-        # Valid pre-release
-        k8s_tags = k8s_tags[1:]
-    # Discard all other pre-releases.
-    return [tag for tag in k8s_tags if not is_stable_release(tag)]
+    seen_stable_minors = set()
+    obsolete = []
+
+    for tag in k8s_tags:
+        if is_stable_release(tag):
+            version = Version(tag.lstrip("v"))
+            seen_stable_minors.add((version.major, version.minor))
+        else:
+            version = Version(tag.lstrip("v").split("-")[0])
+            if (version.major, version.minor) in seen_stable_minors:
+                obsolete.append(tag)
+
+    return obsolete
 
 
-def _exec(cmd: List[str], check=True, timeout=EXEC_TIMEOUT, cwd=None):
+def _exec(
+    cmd: List[str], check=True, capture_output=True, timeout=EXEC_TIMEOUT, cwd=None
+):
```

Comment:
it looks like you just want this to wrap `subprocess.run` to return strings at this point?

```suggestion
def _exec(*args, **kwargs) -> tuple[str, str]:
```